packages/mui-utils/src/resolveProps.ts

```diff
@@ -4,14 +4,43 @@
  * @param {object} props
  * @returns {object} resolved props
  */
-export default function resolveProps<T extends { className?: string } & Record<string, unknown>>(
-  defaultProps: T,
-  props: T,
-) {
+export default function resolveProps<
+  T extends {
+    components?: Record<string, unknown>;
+    componentsProps?: Record<string, unknown>;
+    slots?: Record<string, unknown>;
+    slotProps?: Record<string, unknown>;
+  } & Record<string, unknown>,
+>(defaultProps: T, props: T) {
   const output = { ...props };
 
-  Object.keys(defaultProps).forEach((propName: keyof T) => {
-    if (output[propName] === undefined) {
+  (Object.keys(defaultProps) as Array<keyof T>).forEach((propName) => {
+    if (propName.toString().match(/^(components|slots)$/)) {
+      output[propName] = {
+        ...(defaultProps[propName] as any),
+        ...(output[propName] as any),
+      };
+    } else if (propName.toString().match(/^(componentsProps|slotProps)$/)) {
```

Comment:
Should we also merge props like `TransitionProps` (just by one depth lile `slots` and `components` ?
To solve #34978

Or should it be a distinct PR ?

Reply:
I think it should be a separate PR. I propose to keep it small to fix the main issue.